streams/src/main/java/org/apache/kafka/streams/state/Stores.java

```diff
@@ -0,0 +1,206 @@
+         * @return the key-value store
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context);
+
+        /**
+         * Create a key value store that stores the given type of key and value, that records changes to a Kafka topic, and that
+         * uses the {@link #withTime(Time) configured Time} instance.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keyClass the class for the keys, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @param valueClass the class for the values, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null, or if the {@code keyClass} or {@code valueClass}
+         *             are not one of {@code String.class},
+         *             {@code Integer.class}, {@code Long.class}, or
+         *             {@code byte[].class}
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Class<K> keyClass, Class<V> valueClass);
+
+        /**
+         * Create a key value store that records changes to a Kafka topic and that uses the {@link #withTime(Time) configured
+         * Time} instance and the supplied serializers and deserializers.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keySerializer the serializer for keys; may not be null
+         * @param keyDeserializer the deserializer for keys; may not be null
+         * @param valueSerializer the serializer for values; may not be null
+         * @param valueDeserializer the deserializer for values; may not be null
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Serializer<K> keySerializer, Deserializer<K> keyDeserializer,
+                                                 Serializer<V> valueSerializer, Deserializer<V> valueDeserializer);
+    }
+
+    /**
+     * The factory for creating in-memory key-value stores.
+     */
+    public static interface InMemoryFactory extends KeyValueStoreFactory {
```

Comment:
@guozhangwang, actually, I was thinking that `create(String name, ProcessorContext)` would stay the same and would return the general `StoreFactory` interface (since all `StateStore` instances probably need a name and context), but then `StoreFactory` would have different methods for different kinds of stores. Here's how `StoreFactory` might be enhanced to add some currently-hypothetical stores:

    public interface StoreFactory {
       /* Start of proposed methods for current KeyValueStore implementations*/
       ValueFactory<String> withStringKeys();
       ValueFactory<String> withIntegerKeys();
       ValueFactory<String> withLongKeys();
       ValueFactory<String> withKeys( Class<K> keyClass );
       ValueFactory<String> withKeys( Serializer<K> keySer, Deserializer<K> keyDeser );

       /* Start of hypothetical concepts that could take a variety of forms */

       // document storage ...
       DocumentFactory forDocuments();

       // queryable index storage ...
       IndexFactory usingIndexes( String definition );

       // rdbms ...
       DatabaseFactory withDatabase(...);
    } 

See [my concept](https://gist.github.com/rhauch/ad9a30bb47a041f5301d) for more details on what it'd look like if I changed this commit. I've already verified that this code works and is easily used.

It's a tricky compromise between adding too many levels vs keeping things separated. Right now, `StoreFactory` would contain the various starter methods, while the result of those methods return `ValueFactory`, `DocumentFactory`, `IndexFactory`, `DatabaseFactory`, or something else, and all of those interfaces would completely independent.

    DocumentStore store3 = Stores.create("name", context)
                                 .forDocuments()
                                 .storedCompressed()
                                 .blahBlah()
                                 .build();

If this makes sense and seems useful, my next step would be to add a commit that changes the `Stores` class to use [this new design](https://gist.github.com/rhauch/ad9a30bb47a041f5301d).